cirq/ops/three_qubit_gates.py

```diff
@@ -145,6 +147,126 @@ def __str__(self) -> str:
         return 'CCZ**{}'.format(self._exponent)
 
 
+@value.value_equality()
+class ThreeQubitDiagonalGate(gate_features.ThreeQubitGate):
+    """A gate given by a diagonal 8x8 matrix.
+
+    The list of angles passed to the constructor should be in radians.
+    """
+
+    def __init__(self,
+                 diag_angles_radians: List[Union[float, sympy.Basic]]) -> None:
+        self._diag_angles_radians = diag_angles_radians  # type: List[float]
+
+    def _is_parameterized_(self):
+        return any([
+            isinstance(angle, sympy.Basic)
+            for angle in self._diag_angles_radians
+        ])
+
+    def _has_unitary_(self) -> bool:
+        return not self._is_parameterized_()
+
+    def _unitary_(self) -> np.ndarray:
+        if self._is_parameterized_():
+            return NotImplemented
+        return np.diag(
+            [np.exp(1j * angle) for angle in self._diag_angles_radians])
+
+    def _resolve_parameters_(self, param_resolver: 'cirq.ParamResolver'
+                            ) -> 'ThreeQubitDiagonalGate':
+        return ThreeQubitDiagonalGate([
+            protocols.resolve_parameters(angle, param_resolver)
+            for angle in self._diag_angles_radians
+        ])
+
+    def _circuit_diagram_info_(self, args: protocols.CircuitDiagramInfoArgs
+                              ) -> protocols.CircuitDiagramInfo:
+        return protocols.CircuitDiagramInfo(('diag', '#2', '#3'))
+
+    def __pow__(self, exponent: Any) -> 'ThreeQubitDiagonalGate':
+        return ThreeQubitDiagonalGate([
+            protocols.mul(angle, exponent, NotImplemented)
```

Comment:
This will do the wrong thing when `NotImplemented` is returned. Add a test that confirms `cirq.pow(ThreeQubitDiagonalGate(...), "test", None) is None`.

Reply:
Ah I see, added the test.